lib/logstash/filters/kv.rb

```diff
@@ -292,11 +292,32 @@ def register
     @remove_char_value_re = Regexp.new("[#{@remove_char_value}]") if @remove_char_value
     @remove_char_key_re = Regexp.new("[#{@remove_char_key}]") if @remove_char_key
 
-    valueRxString = "(?:\"([^\"]+)\"|'([^']+)'"
-    valueRxString += "|\\(([^\\)]+)\\)|\\[([^\\]]+)\\]|<([^>]+)>" if @include_brackets
-    valueRxString += "|((?:\\\\ |[^" + @field_split + "])+))"
-    @scan_re = Regexp.new("((?:\\\\ |[^" + @field_split + @value_split + "])+)\s*[" + @value_split + "]\s*" + valueRxString)
-    @value_split_re = /[#{@value_split}]/
+    field_split = /[#{@field_split}]/ # TODO: support multi-char splitters by replacing this regexp
+    value_split = /[#{@value_split}]/ # TODO: support multi-char splitters by replacing this regexp
+
+    value_pattern = begin
+      value_patterns = []
+      value_patterns << /"([^"]+)"/ # quoted double
+      value_patterns << /'([^']+)'/ # quoted single
+      if @include_brackets
+        value_patterns << /\(([^)]+)\)/ # bracketed paren
+        value_patterns << /\[([^\]]+)]/ # bracketed square
+        value_patterns << /<([^>]+)>/   # bracketed angle
+      end
+
+      # an unquoted value is a sequence of characters or escaped spaces before a `field_split`.
+      value_patterns << /((?:\\ |.)+?(?!=#{field_split}))/
```

Comment:
A couple notes about how this expression came to be:

Was effectively:
~~~ ruby
Regexp.new("((?:\\\\ |^" + @field_split + "])+)")
~~~

The following literal regexp, which requires less escaping, is identical:
~~~ ruby
/((?:\\ |[^#{@field_split}])+)/
~~~


The negated-character-class is swapped out (and slightly re-arranged) with a zero-width negative lookahead containing a _positive_-character-class wrapping `@field_split`:
~~~ ruby
/((?:\\ |.)+?(?!=[#{@field_split}]))/
~~~

Finally, with the character-class composition extracted (so we can replace the matcher with an expression that matches more than one character:

~~~ ruby
field_split = /[#{@field_split}]/
/((?:\\ |.)+?(?!=#{field_split}))/
~~~

Reply:
❤️

Reply:
The more I think about it, the more I think I have a bug here -- that's a _negative_ lookahead, so I'm actually not quite sure how the resulting expression is successfully selecting the last character of a value (since the last character _does_ have a field-splitter after it).

There's a pretty big corpus of tests with wildly varying inputs, and they all pass, but the above intuition makes me wary. I'll see if I can prove/disprove this theory shortly.

Below, I took a similar tack and wound up with a _positive_ lookahead, which "feels" more correct.